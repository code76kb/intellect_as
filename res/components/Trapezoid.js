import React, { useState, useEffect } from "react";
import { View } from "react-native";
import PropTypes from 'prop-types'

import { Surface, Shape, Path } from '@react-native-community/art';


// TODO : Min Max window size and StackSize checks

function generateTrapezoidPath(config) {

    // console.log("CONFIG :" + JSON.stringify(config));
    // initial points 
    let X = config.X;
    let Y = config.Y;
    let points = [];

    const offset = config.radius; // 8
    const R = 0 // X, Y radius 

    points.push({ x: X + config.width, y: Y });
    points.push({ x: X + config.width - offset, y: Y + config.height });
    points.push({ x: X + offset, y: Y + config.height });

    var path = new Path();

    path = path.moveTo(X, Y + config.radius);
    path = path.arcTo(X + config.radius, Y, R, R)

    path = path.lineTo(points[0].x - config.radius, points[0].y);
    path = path.arcTo(points[0].x, points[0].y + config.radius, R, R)

    path = path.lineTo(points[1].x, points[1].y - config.radius);
    path = path.arcTo(points[1].x - config.radius, points[1].y, R, R)

    path = path.lineTo(points[2].x + config.radius, points[2].y);
    path = path.arcTo(points[2].x, points[2].y - config.radius, R, R)

    path = path.lineTo(X, Y + config.radius);

    return path;
}

function generateTrapezoidStack(props) {
    const squashFactor = 2;
    const paddingFactor = 5; // TODO expose on props
    const curveRadius = (props.height / (props.stackSize * 7)); // factor:7
    const paddingVertical = props.height / (props.stackSize * paddingFactor);
    const stackSize = props.stackSize;
    const stackWidth = props.width;
    const stackHeight = props.height;
    const widthDelta = stackWidth / (stackSize * 2);

    const config = {
        height: (stackHeight - (stackSize * paddingVertical)) / stackSize,
        width: 0,
        radius: curveRadius,
        noOfPoints: 4,
        X: 0,
        Y: 0
    }
    let paths = []
    let configs = []
    for (let i = 0; i < stackSize; i++) {
        let conf = { ...config };
        conf.Y = i * (config.height + paddingVertical);
        conf.X = config.X + ((i * squashFactor) * (widthDelta / 2))
        conf.width = stackWidth - ((i * squashFactor) * (widthDelta))
        configs.push(conf);
        paths.push(generateTrapezoidPath(conf));
    }
    return { paths, configs };
}

export default function Trapezoid(props) {
    const { paths, configs } = generateTrapezoidStack(props);
    const [progress, setProgress] = useState(props.stackSize - props.value);

    useEffect(() => {
        if (progress != (props.stackSize - props.value)){
            setProgress(props.stackSize - props.value)
        }
    }, [props.value])

    function getValue() {
        return progress //props.stackSize - props.value;
    }

    function onTouch(e) {
        for (let i = 0; i < configs.length; i++) {
            if (e.nativeEvent.locationY < configs[i].Y + configs[i].height) {
                setProgress(i);
                props.onChange(props.stackSize - i)
                return;
            }
        }
        if (progress == props.stackSize - 1)
            props.onChange(1);
    }


    return (
        <View style={{ width: props.width, height: props.height }} onTouchStart={onTouch}>
            <Surface width={props.width} height={props.height}>
                {
                    paths.map((path, index) => {
                        return <Shape key={`trap_${index}`} d={path} stroke="white" fill={getValue() <= index ? 'white' : "#819da7"} strokeWidth={1} />
                    })
                }

            </Surface>
        </View>
    );
}



Trapezoid.propTypes = {
    height: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    stackSize: PropTypes.number.isRequired,
    value: PropTypes.number,
    onChange: PropTypes.func
}

Trapezoid.defaultProps = {
    height: 400,
    width: 250,
    stackSize: 5,
    value: 1
}
