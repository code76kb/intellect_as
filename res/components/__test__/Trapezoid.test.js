import React from 'react';
import renderer from 'react-test-renderer';
import Trapezoid from '../Trapezoid' 
it('Render Trapezoid  Comp :', () => {
    const tree = renderer.create(
        <Trapezoid
        height={300}
        width={240}
        stackSize={5}
        value={2}
        onChange={()=>{}}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
})