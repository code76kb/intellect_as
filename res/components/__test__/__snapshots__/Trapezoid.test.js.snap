// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Trapezoid  Comp : 1`] = `
<View
  onTouchStart={[Function]}
  style={
    Object {
      "height": 300,
      "width": 240,
    }
  }
>
  <ARTSurfaceView
    style={
      Array [
        undefined,
        Object {
          "height": 300,
          "width": 240,
        },
      ]
    }
  >
    <ARTShape
      d={
        Array [
          0,
          0,
          8.571428571428571,
          4,
          8.571428571428571,
          8.571428571428571,
          8.571428571428571,
          3.141592653589793,
          -1.5707963267948966,
          1,
          2,
          231.42857142857142,
          0,
          3,
          236.16244071283538,
          0,
          240,
          3.8375592871646287,
          240,
          8.571428571428571,
          2,
          231.42857142857142,
          39.42857142857143,
          3,
          231.42857142857142,
          44.16244071283537,
          227.5910121414068,
          48,
          222.85714285714283,
          48,
          2,
          17.142857142857142,
          48,
          3,
          12.4089878585932,
          48,
          8.571428571428573,
          44.16244071283538,
          8.571428571428571,
          39.42857142857143,
          2,
          0,
          8.571428571428571,
        ]
      }
      fill={
        Array [
          0,
          0.5058823529411764,
          0.615686274509804,
          0.6549019607843137,
          1,
        ]
      }
      opacity={1}
      stroke="#ffffff"
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
    <ARTShape
      d={
        Array [
          0,
          24,
          68.57142857142857,
          4,
          32.57142857142857,
          68.57142857142857,
          8.57142857142857,
          3.141592653589793,
          -1.5707963267948966,
          1,
          2,
          207.42857142857142,
          60,
          3,
          209.70185562719607,
          60,
          211.88203202347296,
          60.90305863170358,
          213.4894866958847,
          62.51051330411531,
          3,
          215.09694136829643,
          64.11796797652703,
          216,
          66.2981443728039,
          216,
          68.57142857142857,
          2,
          207.42857142857142,
          99.42857142857143,
          3,
          207.42857142857142,
          104.16244071283538,
          203.5910121414068,
          108,
          198.85714285714283,
          108,
          2,
          41.14285714285714,
          108,
          4,
          41.14285714285714,
          99.42857142857143,
          8.57142857142857,
          1.5707963267948966,
          3.141592653589793,
          1,
          2,
          24,
          68.57142857142857,
        ]
      }
      fill={
        Array [
          0,
          0.5058823529411764,
          0.615686274509804,
          0.6549019607843137,
          1,
        ]
      }
      opacity={1}
      stroke="#ffffff"
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
    <ARTShape
      d={
        Array [
          0,
          48,
          128.57142857142858,
          3,
          48,
          123.83755928716464,
          51.83755928716462,
          120,
          56.57142857142857,
          120,
          2,
          183.42857142857142,
          120,
          4,
          183.42857142857142,
          128.57142857142858,
          8.571428571428584,
          -1.5707963267948966,
          0,
          1,
          2,
          183.42857142857142,
          159.42857142857142,
          4,
          174.85714285714283,
          159.42857142857142,
          8.571428571428584,
          -0,
          1.5707963267948966,
          1,
          2,
          65.14285714285714,
          168,
          3,
          60.4089878585932,
          168,
          56.57142857142857,
          164.16244071283538,
          56.57142857142857,
          159.42857142857142,
          2,
          48,
          128.57142857142858,
        ]
      }
      fill={
        Array [
          0,
          0.5058823529411764,
          0.615686274509804,
          0.6549019607843137,
          1,
        ]
      }
      opacity={1}
      stroke="#ffffff"
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
    <ARTShape
      d={
        Array [
          0,
          72,
          188.57142857142858,
          3,
          72,
          183.83755928716462,
          75.83755928716462,
          180,
          80.57142857142857,
          180,
          2,
          159.42857142857142,
          180,
          4,
          159.42857142857142,
          188.57142857142858,
          8.571428571428584,
          -1.5707963267948966,
          0,
          1,
          2,
          159.42857142857142,
          219.42857142857142,
          4,
          150.85714285714283,
          219.42857142857142,
          8.571428571428584,
          -0,
          1.5707963267948966,
          1,
          2,
          89.14285714285714,
          228,
          3,
          84.40898785859319,
          228,
          80.57142857142857,
          224.16244071283538,
          80.57142857142857,
          219.42857142857142,
          2,
          72,
          188.57142857142858,
        ]
      }
      fill={
        Array [
          0,
          1,
          1,
          1,
          1,
        ]
      }
      opacity={1}
      stroke="#ffffff"
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
    <ARTShape
      d={
        Array [
          0,
          96,
          248.57142857142858,
          3,
          96,
          243.83755928716462,
          99.83755928716462,
          240,
          104.57142857142857,
          240,
          2,
          135.42857142857142,
          240,
          4,
          135.42857142857142,
          248.57142857142858,
          8.571428571428584,
          -1.5707963267948966,
          0,
          1,
          2,
          135.42857142857142,
          279.42857142857144,
          3,
          135.42857142857142,
          284.1624407128354,
          131.5910121414068,
          288,
          126.85714285714285,
          288,
          2,
          113.14285714285714,
          288,
          3,
          108.40898785859319,
          288,
          104.57142857142857,
          284.1624407128354,
          104.57142857142857,
          279.42857142857144,
          2,
          96,
          248.57142857142858,
        ]
      }
      fill={
        Array [
          0,
          1,
          1,
          1,
          1,
        ]
      }
      opacity={1}
      stroke="#ffffff"
      strokeCap={1}
      strokeDash={null}
      strokeJoin={1}
      strokeWidth={1}
      transform={
        Array [
          1,
          0,
          0,
          1,
          0,
          0,
        ]
      }
    />
  </ARTSurfaceView>
</View>
`;
